File: movie-ticket-system/movie-service/app/services/movie_service.py
from app.config.database import movies_collection
from bson import ObjectId
from datetime import datetime, date


def _serialize_movie(movie):
    if not movie:
        return None
    doc = dict(movie)
    doc["_id"] = str(doc["_id"])
    if "hallId" in doc and doc["hallId"]:
        doc["hallId"] = str(doc["hallId"])
    for k in ["startDate", "endDate"]:
        if k in doc and doc[k]:
            doc[k] = doc[k].isoformat() if hasattr(doc[k], "isoformat") else str(doc[k])
    return doc


def create_movie(movie: dict) -> str:
    data = {**movie}
    if "price" in data and data["price"]:
        data["price"] = dict(data["price"])
    else:
        data["price"] = {"ODC_Full": 0, "ODC_Half": 0, "Balcony": 0, "Box": 0}
    if isinstance(data.get("hallId"), str):
        data["hallId"] = ObjectId(data["hallId"])
    for f in ["startDate", "endDate"]:
        if f in data:
            v = data[f]
            if isinstance(v, str):
                data[f] = datetime.fromisoformat(v.replace("Z", "+00:00"))
            elif isinstance(v, date) and not isinstance(v, datetime):
                data[f] = datetime.combine(v, datetime.min.time())
    result = movies_collection.insert_one(data)
    return str(result.inserted_id)


def get_all_movies(hall_id: str = None):
    filter_query = {}
    if hall_id:
        filter_query["hallId"] = ObjectId(hall_id)
    movies = []
    # latest created first by ObjectId timestamp
    for movie in movies_collection.find(filter_query).sort("_id", -1):
        movies.append(_serialize_movie(movie))
    return movies


def get_movie(movie_id: str):
    movie = movies_collection.find_one({"_id": ObjectId(movie_id)})
    return _serialize_movie(movie)


def update_movie(movie_id: str, data: dict):
    clean = {k: v for k, v in data.items() if v is not None}
    if "hallId" in clean and isinstance(clean["hallId"], str):
        clean["hallId"] = ObjectId(clean["hallId"])
    for f in ["startDate", "endDate"]:
        if f in clean:
            v = clean[f]
            if isinstance(v, str):
                clean[f] = datetime.fromisoformat(v.replace("Z", "+00:00"))
            elif isinstance(v, date) and not isinstance(v, datetime):
                clean[f] = datetime.combine(v, datetime.min.time())
    if "price" in clean and clean["price"]:
        clean["price"] = dict(clean["price"])
    movies_collection.update_one({"_id": ObjectId(movie_id)}, {"$set": clean})


def delete_movie(movie_id: str):
    movies_collection.delete_one({"_id": ObjectId(movie_id)})

File: movie-ticket-system/movie-service/src/services/movie.service.js
const mongoose = require("mongoose");
const Movie = require("../models/movie.model");

function normalizePrice(price) {
  if (!price) {
    return {
      ODC_Full: 0,
      ODC_Half: 0,
      Balcony: 0,
      Box: 0,
    };
  }
  return {
    ODC_Full: Number(price.ODC_Full || 0),
    ODC_Half: Number(price.ODC_Half || 0),
    Balcony: Number(price.Balcony || 0),
    Box: Number(price.Box || 0),
  };
}

async function createMovie(payload) {
  const data = { ...payload };

  if (!data.hallId) {
    throw new Error("hallId is required");
  }

  data.hallId = new mongoose.Types.ObjectId(data.hallId);
  data.price = normalizePrice(data.price);

  const movie = await Movie.create(data);
  return String(movie._id);
}

async function getAllMovies(hallId) {
  const filter = {};

  if (hallId) {
    filter.hallId = new mongoose.Types.ObjectId(hallId);
  }

  const movies = await Movie.find(filter).sort({ _id: -1 }).lean();
  return movies.map((m) => ({
    ...m,
    _id: String(m._id),
    hallId: String(m.hallId),
  }));
}

async function getMovieById(id) {
  if (!mongoose.Types.ObjectId.isValid(id)) {
    return null;
  }

  const movie = await Movie.findById(id).lean();
  if (!movie) return null;

  return {
    ...movie,
    _id: String(movie._id),
    hallId: String(movie.hallId),
  };
}

async function updateMovieById(id, payload) {
  if (!mongoose.Types.ObjectId.isValid(id)) {
    return null;
  }

  const data = {};
  Object.entries(payload).forEach(([key, value]) => {
    if (value !== null && value !== undefined) {
      data[key] = value;
    }
  });

  if (data.hallId) {
    data.hallId = new mongoose.Types.ObjectId(data.hallId);
  }

  if (data.price) {
    data.price = normalizePrice(data.price);
  }

  const movie = await Movie.findByIdAndUpdate(id, data, {
    new: true,
  }).lean();

  if (!movie) return null;

  return {
    ...movie,
    _id: String(movie._id),
    hallId: String(movie.hallId),
  };
}

async function deleteMovieById(id) {
  if (!mongoose.Types.ObjectId.isValid(id)) {
    return null;
  }

  const result = await Movie.findByIdAndDelete(id).lean();
  if (!result) return null;
  return true;
}

module.exports = {
  createMovie,
  getAllMovies,
  getMovieById,
  updateMovieById,
  deleteMovieById,
};

